### ISSUE TITLE
Ginkgo timed out waiting for all parallel nodes to report back!

### ISSUE THREAD

**Issue by cevich** [containers/libpod#3114]
/kind bug

**Description**

Various Cirrus-CI tasks fail with this text after about 20-minutes of runtime:

```
[+1346s] 	 -------------------------------------------------------------------
[+1346s] 	|                                                                   |
[+1346s] 	|  Ginkgo timed out waiting for all parallel nodes to report back!  |
[+1346s] 	|                                                                   |
[+1346s] 	 -------------------------------------------------------------------
```

However, the ginkgo timeout should be set @ 90-minutes.  Emergence of this behavior seems to coincide with wide-spread flakes we're currently seeing following PR #3072 7b54ebb48fecb4205304586a57a3e6b9b96befa0

**Steps to reproduce the issue:**

1. Unknown, seems to happen randomly

**Describe the results you received:**

See description

**Describe the results you expected:**

All CI tasks should pass

**Additional information you deem important (e.g. issue happens only occasionally):**

Poking around on google, this error can happen due to bad signal handling between ginkgo and the test subject.  Many issues were filed against ginkgo and other projects eluding to the difficulty of debugging this situation.  Found some concluding with an already merged ginkgo fix, mentioning the use of `-debug` option.

**Comment by mheon** [comment]
This has to be ginkgo itself. Possibly a separate timeout for parallel testing we missed?

**Comment by cevich** [comment]
From #3083
> Another timeout flake. It seems like it's consistently 30 minutes - but we're setting 90. This seems like it might be an infra thing?

@mheon I don't think this is related to infrastructure, though I would not rule that out yet.  All the ginkgo-related issues I trolled with this error message, revolved around a test running in a ginkgo node missing/mishandling a signal from ginkgo (STOP I believe).  There were also TONS of complaints and proposed fixes to ginkgo since people found this incredibly difficult to debug.

Though since #3083 happened _after_ your fix to the 'pod rm' / cgroups flake, that means we can probably rule rule "correlation != causation".  In other words, this new timeout thing is something different/unrelated.

I would like to try two things:

1. Remove `-nodes 3` and run tests repeatedly to see if any flake
2. Add `-debug` (with `-nodes 3`) repeatedly, to see if if timeout-flakes, and then gives us more detail of what's happening.

**Comment by cevich** [comment]
> This has to be ginkgo itself. Possibly a separate timeout for parallel testing we missed?

Could be, but from my trolling the error, its seems likely a ginkgo thread (node) that's "hanging" because a test is hanging.

**Comment by cevich** [comment]
TEST PR: #3115

**Comment by mheon** [comment]
@cevich I think we're hitting some sort of internal Ginkgo PR (inconsistently). What's strange is that it seems like 30 minutes of *overall* time on the VM, then it stops, consistently - not 30 minutes in the Ginkgo integration tests

**Comment by mheon** [comment]
Internal Ginkgo Timeout. I need more coffee.

**Comment by cevich** [comment]
A/B testing in 3115 finished with interesting results:

* When `-nodes3` is removed (executes all test in serial) the same test failed every time:  `podman run notify_socket`
* One of the runs with `-debug` enabled + `-nodes 3` also seems to have flaked on this test.
* The other two runs with `-debug` enabled passed fine.

So the instrumented testing results seem useful and mirror the randomness we see in the wild.  Digging into the `-debug` enabled test run of `podman run notify_socket` to see if it looks like the failure output from the serial runs...

**Comment by cevich** [comment]
...common error across all failures seems to be:

```
Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test334660585/crio --runroot /tmp/podman_test334660585/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test334660585 --storage-driver vfs stop -a --timeout 0
[+2337s] Error: timed out waiting for the condition
```

Is that correct, it should use `--timeout 0`, or a typo?

**Comment by mheon** [comment]
The `--timeout 0` for `podman stop` indicates that it shouldn't bother with `SIGINT`; just go straight to `SIGKILL` immediately.

**Comment by mheon** [comment]
It's timing out waiting for the exit file to be created, I think. Your error is almost certainly coming out of `waitForExitFileAndSync()` in `container_internal.go`.

**Comment by cevich** [comment]
It's curious how this test could hang then, SIGKILL is not trapable.  Looking at the tracebacks from ginkgo, I see no difference between the `-debug` (paralel) and the `-nodes 1` (serial) failure cases.  But somehow, when running under `-nodes 3` this test _sometimes_ becomes unresponsive to ginkgo trying to signal it.

**Comment by cevich** [comment]
``[cevich@cevich libpod]$ diff -Naur A B``

```diff
--- A	2019-05-13 14:03:34.467347559 -0400
+++ B	2019-05-13 14:03:32.359384378 -0400
@@ -1,33 +1,31 @@
 Podman run 
   podman run notify_socket
   /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:275
-
 [BeforeEach] Podman run
   /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:27
 Restoring docker.io/library/alpine:latest...
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs load -q -i /tmp/alpine-latest.tar
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs load -q -i /tmp/alpine-latest.tar
 Loaded image(s): docker.io/library/alpine:latest
 Restoring docker.io/library/busybox:latest...
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs load -q -i /tmp/busybox-latest.tar
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs load -q -i /tmp/busybox-latest.tar
 Loaded image(s): docker.io/library/busybox:latest
 [It] podman run notify_socket
   /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:275
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs run docker.io/library/alpine:latest printenv NOTIFY_SOCKET
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs run docker.io/library/alpine:latest printenv NOTIFY_SOCKET
 /run/notify/notify.sock
 [AfterEach] Podman run
   /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:37
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs stop -a --timeout 0
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs stop -a --timeout 0
 Error: timed out waiting for the condition
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs pod stop -a -t 0
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs pod stop -a -t 0
 output: 
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs pod rm -fa
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs pod rm -fa
 output: 
-Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test527522307/crio --runroot /tmp/podman_test527522307/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test527522307 --storage-driver vfs rm -fa
-time="2019-05-13T13:07:28-04:00" level=error msg="cleanup storage: failed to cleanup container 77a25a509b084ad456b4fff852566ffbd5884c93ebe005316402f0f68c7bbff8 storage: error unmounting container 77a25a509b084ad456b4fff852566ffbd5884c93ebe005316402f0f68c7bbff8 root filesystem: layer is not mounted" 
-Error: error cleaning up container 77a25a509b084ad456b4fff852566ffbd5884c93ebe005316402f0f68c7bbff8: error unmounting container 77a25a509b084ad456b4fff852566ffbd5884c93ebe005316402f0f68c7bbff8 storage: error unmounting container 77a25a509b084ad456b4fff852566ffbd5884c93ebe005316402f0f68c7bbff8 root filesystem: layer is not mounted
-
+Running: /var/tmp/go/src/github.com/containers/libpod/bin/podman --root /tmp/podman_test674115587/crio --runroot /tmp/podman_test674115587/crio-run --runtime /usr/bin/runc --conmon /usr/libexec/podman/conmon --cni-config-dir /etc/cni/net.d --cgroup-manager cgroupfs --tmpdir /tmp/podman_test674115587 --storage-driver vfs rm -fa
+time="2019-05-13T13:21:46-04:00" level=error msg="cleanup storage: failed to cleanup container 0302b76fca15c64ae63f08888176b526e565e51e542048d337bd954253191fd6 storage: error unmounting container 0302b76fca15c64ae63f08888176b526e565e51e542048d337bd954253191fd6 root filesystem: layer is not mounted" 
+Error: error cleaning up container 0302b76fca15c64ae63f08888176b526e565e51e542048d337bd954253191fd6: error unmounting container 0302b76fca15c64ae63f08888176b526e565e51e542048d337bd954253191fd6 storage: error unmounting container 0302b76fca15c64ae63f08888176b526e565e51e542048d337bd954253191fd6 root filesystem: layer is not mounted
 
-â€¢ Failure [96.579 seconds]
+â€¢ Failure [96.313 seconds]
 Podman run
 /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:20
   podman run notify_socket [It]
@@ -40,15 +38,15 @@
 
   Full Stack Trace
   	/var/tmp/go/src/github.com/containers/libpod/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go:49 +0x81
-  github.com/containers/libpod/vendor/github.com/onsi/gomega/gexec.(*Session).Wait(0xc000933020, 0xc000c61118, 0x1, 0x1, 0x1)
+  github.com/containers/libpod/vendor/github.com/onsi/gomega/gexec.(*Session).Wait(0xc001e65290, 0xc00103b118, 0x1, 0x1, 0x1)
   	/var/tmp/go/src/github.com/containers/libpod/vendor/github.com/onsi/gomega/gexec/session.go:142 +0xc8
-  github.com/containers/libpod/test/utils.(*PodmanSession).WaitWithDefaultTimeout(0xc000853590)
+  github.com/containers/libpod/test/utils.(*PodmanSession).WaitWithDefaultTimeout(0xc002180f28)
   	/var/tmp/go/src/github.com/containers/libpod/test/utils/utils.go:317 +0x89
   github.com/containers/libpod/test/e2e.glob..func70.23()
   	/var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:295 +0x406
-  github.com/containers/libpod/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc000de31a0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
+  github.com/containers/libpod/vendor/github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync(0xc001057380, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
   	/var/tmp/go/src/github.com/containers/libpod/test/e2e/common_test.go:87 +0xe0
-  testing.tRunner(0xc000120a00, 0x13739b8)
+  testing.tRunner(0xc000268500, 0x13739b8)
   	/usr/lib/golang/src/testing/testing.go:827 +0xbf
   created by testing.(*T).Run
   	/usr/lib/golang/src/testing/testing.go:878 +0x35c
```

**Comment by cevich** [comment]
ya...only thing that changes appears to be memory addresses and names/locations we would expect to differ.  All the line numbers and other significant details remain identical between serial and parallel failures.

**Comment by mheon** [comment]
I think we are seeing the container die, but we're not finding the exit file containing details about how the container exited (with what code, when). Conmon could be dying prematurely, potentially? Or the file is created in a place we don't know to look?

**Comment by cevich** [comment]
Hmmm, I don't know about most of this stuff, but what sticks out to me is:  The test uses a number of `defer` statements to cleanup a socket and it's file handle.  However, the failure trace happens during 'AfterEach`.  Since this happens randomly, perhaps the file-handle defer and socket close defer are racing eachother, or (if possible) racing AfterEach()?

**Comment by cevich** [comment]
Experimenting with fixing the cleanup order of the socket, fd, and env var...

**Comment by cevich** [comment]
@mheon assuming my experiment will fail, and you're right:  How can I verify ``we're not finding the exit file containing details``?  Would running an `ls -la` during `AfterEach` suffice?  What directory should it list?

**Comment by mheon** [comment]
Exit files are going to be in libpod's temporary directory (`--tmpdir`), in a subdirectory called `exit`, with their name being the full ID of the container in question

**Comment by cevich** [comment]
mmmm, dang.  That sounds like it would be difficult to find from AfterEach :disappointed: 

Maybe I could temporarily move the `AfterEach` code into this specific test, where it could have ready access to those details.  Then it would be a simple matter of spamming the test until it fails with proof (or not) of a missing(?) exit file.

**Comment by cevich** [comment]
@mheon okay, well my experiment did fail on all 10 attempts, we still get:

```
[+1774s] â€¢ Failure [96.356 seconds]
[+1774s] Podman run
[+1774s] /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:20
[+1774s]   podman run notify_socket [It]
[+1774s]   /var/tmp/go/src/github.com/containers/libpod/test/e2e/run_test.go:275
[+1774s] 
[+1774s]   Timed out after 90.000s.
[+1774s]   Expected process to exit.  It did not.
[+1774s] 
[+1774s]   /var/tmp/go/src/github.com/containers/libpod/test/utils/utils.go:317
```

Where `test/utils/utils.go:317` is the first line of code in the function:

```golang
// WaitWithDefaultTimeout waits for process finished with defaultWaitTimeout
func (s *PodmanSession) WaitWithDefaultTimeout() {
    s.Wait(defaultWaitTimeout)
    os.Stdout.Sync()
    os.Stderr.Sync()
    fmt.Println("output:", s.OutputToString())
}
```

**Comment by cevich** [comment]
Removed my experimental fix, and re-worked testing PR to only run this single test x10 with 10x flake attempts each.  This should make it fail fast, and often.  Then I'll see about this exit-file business...

**Comment by cevich** [comment]
@mheon playing around in 1.2.0, while a container is running I see no matching file in `/var/run/libpod/exits/$CID`.  After the container exits, but before I remove it, I see:
``/var/run/libpod/exits/ad53df0c...f64b3-old``.

So in the case of this test, after the normal test code runs, I should NOT find a ``--tmpdir`` file called ``${CID}.old`` (presumably, because the container has not stopped)?

**Comment by cevich** [comment]
(--force pushed: removed 10x flake-attempts, taking too long to fail 90-seconds each)

**Comment by mheon** [comment]
OK, `-old` makes some sense - when we read the file in, we rename it (I forget exactly where) so we won't have conflicts if we restart the container

**Comment by cevich** [comment]
Interesting, so there's actually a double-failure happening here.  The first (test-fail) @baude thinks is a multi-thread collision on the socket name.  The test then goes into AfterEach and that's where we get the hang and timeout.  This seems to be masking the original failure, so I'm not sure what that is.  Trying Brent's socket name fix...

**Comment by cevich** [comment]
@baude with your patch there is no change:  The `printenv NOTIFY_SOCKET` still outputs  ``/run/notify/notify.sock``, but then we hang in `session.WaitWithDefaultTimeout()`.  So `AfterEach` takes over, and fails again.   I double-checked a prior run, to make sure, it's the same behavior :disappointed: 

@mheon I'm assuming there's some "magic" involved with passing NOTIFY_SOCKET from the host to the container.  Could that be setting up the 'podman run' somehow such that it races when exiting?

**Comment by cevich** [comment]
This should be fixed by #3127

**Comment by cevich** [comment]
Sadly reopening this, seems not all the gremlins have been shaken loose :disappointed:
